app/models/user.rb

```diff
@@ -40,6 +33,8 @@ class User < ApplicationRecord
 
   mount_uploader :image, UserUploader
 
+  devise :database_authenticatable, :recoverable, :rememberable, :validatable
```

Comment:
nêu chức năng từng module, custome valdate theo ý mình thì làm thế nào, thử luôn trong pull nhé

Reply:
**database_authenticatable**: Module này chịu trách nhiệm cho việc hash password và xác thực khi người dùng khi đăng nhập. Module này cung cấp method password= để hash password và lưu nó vào cột encrypted_password  và bỏ qua bất kì cột password đã tồn tại trước đó.

**recoverable** : Module này chịu trách nhiệm cho chức năng reset password và gửi hướng dẫn reset password. 
Module này thêm vào devise_for những option sau:

- +reset_password_keys+: Key mà bạn muốn sử dụng khi thực hiện việc recover password cho một tài khoản.
- +reset_password_within+: Thiết lập thời gian cho việc reset password.
- +sign_in_after_reset_password+: Sau khi reset password xong thì có tự động đăng nhập cho user đó không.

**rememberable**: Module quản lý việc tạo và xóa token cho việc ghi nhớ user từ một cookie được lưu. Module này cũng có những method để mã hóa và đưa vào cookie cũng như việc xóa nó trong cookie, cố gắng tìm kiếm record trong DB dựa vào những thông tin được lưu trong cookie. 

**validatable**: Module cung cấp toàn bộ những validate cho email và password. Tự động validate nếu có email, unique và định dạng của email. 
Module này thêm vào devise_for những option sau:
- +email_regexp+: regex cho email 
- +password_length+: Khoảng độ dài của password. Mặc định là 6..128.